File: kits/ts/simple/lux/Position.ts
import {GAME_CONSTANTS} from "./game_constants";
const {DIRECTIONS} = GAME_CONSTANTS;

export type ArrayPosition = [x: number, y: number];

export class Position {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  public isAdjacent(pos: Position): boolean {
    const dx = this.x - pos.x;
    const dy = this.y - pos.y;
    return Math.abs(dx) + Math.abs(dy) <= 1;
  }

  public equals(pos: Position): boolean {
    return this.x === pos.x && this.y === pos.y;
  }

  public translate(direction: string, units: number): Position {
    switch (direction) {
      case DIRECTIONS.NORTH:
        return new Position(this.x, this.y - units);
      case DIRECTIONS.EAST:
        return new Position(this.x + units, this.y);
      case DIRECTIONS.SOUTH:
        return new Position(this.x, this.y + units);
      case DIRECTIONS.WEST:
        return new Position(this.x - units, this.y);
      case DIRECTIONS.CENTER:
        return new Position(this.x, this.y);
    }
  }

  /** Returns distance to pos from this position */
  public distanceTo(pos: Position): number {
    const dx = pos.x - this.x;
    const dy = pos.y - this.y;
    return Math.sqrt(dx * dx + dy * dy);
  }

  /** Returns closest direction to targetPos, or null if staying put is best */
  public directionTo(targetPos: Position): string {
    const checkDirections = [
      DIRECTIONS.NORTH,
      DIRECTIONS.EAST,
      DIRECTIONS.SOUTH,
      DIRECTIONS.WEST,
    ];
    let closestDirection = DIRECTIONS.CENTER;
    let closestDist = this.distanceTo(targetPos);
    checkDirections.forEach((dir) => {
      const newpos = this.translate(dir, 1);
      const dist = targetPos.distanceTo(newpos);
      if (dist < closestDist) {
        closestDist = dist;
        closestDirection = dir;
      }
    });
    return closestDirection;
  }
}

File: kits/java/simple/lux/Position.java
package lux;
public class Position {
  public int x;
  public int y;

  public Position(int x, int y) {
    this.x = x;
    this.y = y;
  }

  public boolean isAdjacent(Position pos) {
    int dx = this.x - pos.x;
    int dy = this.y - pos.y;
    if (Math.abs(dx) + Math.abs(dy) > 1) {
      return false;
    }
    return true;
  }

  public boolean equals(Position pos) {
    return this.x == pos.x && this.y == pos.y;
  }

  public Position translate(Direction direction, int units) {
    switch (direction) {
      case NORTH:
        return new Position(this.x, this.y - units);
      case EAST:
        return new Position(this.x + units, this.y);
      case SOUTH:
        return new Position(this.x, this.y + units);
      case WEST:
        return new Position(this.x - units, this.y);
      case CENTER:
        return new Position(this.x, this.y);
    }
    throw new Error("Did not supply valid direction");
  }

  public double distanceTo(Position pos) {
    return Math.abs(pos.x - this.x) + Math.abs(pos.y - this.y);
  }


  public Direction directionTo(Position targetPos) {
    Direction[] checkDirections = { Direction.NORTH, Direction.EAST, Direction.SOUTH, Direction.WEST, };
    Direction closestDirection = Direction.CENTER;
    double closestDist = this.distanceTo(targetPos);
    for (Direction dir : checkDirections) {
      Position newpos = this.translate(dir, 1);
      double dist = targetPos.distanceTo(newpos);
      if (dist < closestDist) {
        closestDist = dist;
        closestDirection = dir;
      }
    }
    return closestDirection;
  }

  @Override
  public String toString() {
    return "(" + this.x + ", " + this.y + ")";
  }
}
